src/pandas/types/numeric.h

```diff
@@ -44,14 +106,153 @@ class PANDAS_EXPORT IntegerArray : public NumericArray<TYPE> {
 
   bool owns_data() const override;
 
+  Status EnsureMutable();
+
+  // In the case where there are any nulls in this array,
+  // this function will be faster than calling GetNullCount() > 0.
+  bool HasNulls() const {
+    for (auto ii = 0; ii < this->length_; ++ii) {
+      if (IsNull(ii)) { return true; }
+    }
+    return false;
+  }
+
+  // Marks the value at the specified index as null. A pre-condition
+  // is that the valid_bits for this object have already been
+  // initialized.
+  void SetNull(int64_t i) {
+    // Pre-condition is that valid_bits_ is non-null because
+    // operations that would set them (e.g. - operator+=)
+    // would first create the valid_bits_ if needed
+    BitUtil::SetBit(
+        static_cast<MutableBuffer&>(*valid_bits_).mutable_data(), this->offset_ + i);
+  }
+
+  // Marks the value at the specified index as null. A pre-condition
+  // is that the valid_bits for this object have already been
+  // initialized.
+  void SetValid(int64_t i) {
+    // Pre-condition is that valid_bits_ is non-null because
+    // operations that would set them (e.g. - operator+=)
+    // would first create the valid_bits_ if needed
+    BitUtil::ClearBit(
+        static_cast<MutableBuffer&>(*valid_bits_).mutable_data(), this->offset_ + i);
+  }
+
+  // Returns whether or not the value at the specified index has
+  // been marked as null.
+  bool IsNull(int64_t i) const {
+    if (valid_bits_) {
+      // TODO: range check?
+      return BitUtil::GetBit(valid_bits_->data(), this->offset_ + i);
+    } else {
+      return false;
+    }
+  }
+
+  // Copy the list of nulls (aka valid bits) to a buffer
+  Status CopyNulls(int64_t length, std::shared_ptr<Buffer>* out) const {
+    // Pre-condition is that we already have valid_bits_ set
+    return valid_bits_->Copy(this->offset_, BitUtil::BytesForBits(length), out);
+  }
+
+  template <typename OTHER_TYPE>
+  IntegerArray<TYPE>& operator+=(const IntegerArray<OTHER_TYPE>& other) {
+    // TODO: check Status and throw exception
+    EnsureMutable();
+    auto length = std::min(this->length_, other.length());
```

Comment:
Thinking out loud here. I have been thinking about a slightly different approach with implementing operators in generality, by instead unpacking the containers into simpler data structures, like:

```c++
# SFINAE
template <typename T, typename ENABLED = void>
struct TypeData {};

template <typename T>
struct TypeData<T, std::enable_if<std::is_integer<T>, T>::type> {
  T* data;
  uint8_t* valid_bits;
  int64_t length;
  int64_t offset;
}
```

and then have arithmetic methods

```c++
template <typename LEFT_TYPE, typename RIGHT_TYPE, typename OUT_TYPE>
void add(TypeData<LEFT_TYPE> left, TypeData<RIGHT_TYPE> right, TypeData<OUT_TYPE> out) {
  ...
  for (int64_t i = 0; i < left.length; ++i) {
    out.data[i] = left.data[i] + right.data[i];
  } 
  ...
}
```

In the case of inplace-add, in the operator dispatcher, you would have::

```
TypeData<LEFT_TYPE> left(...);
TypeData<RIGHT_TYPE> right(...);
add(left, right, left);
```

You would have to deal with whether to allocate the validity/null bitmap prior to invoking this function, but I think it's probably better to handle the mutability checks / memory allocation / etc. separate from the analytics. Then you can box the output as you wish.

Of course, you need to have a forward declaration someplace

```c++
extern template <typename LEFT_TYPE, typename RIGHT_TYPE, typename OUT_TYPE>
void add(TypeData<LEFT_TYPE> left, TypeData<RIGHT_TYPE> right, TypeData<OUT_TYPE> out);
```

and then instantiate the matrix of types you're choosing to support. Some analytical kernels might need to allocate memory based on what's in the data, and so we could do something like

```c++
void kernel_name(OperatorContext* ctx, TypeData<ARG1_TYPE> arg1, ...);
```

I'd also like to maximize code sharing amongst related operator implementations. 

I'm going to take a crack at implementing part of `Take` under this regime (see https://github.com/pandas-dev/pandas/blob/master/pandas/src/algos_take_helper.pxi). Will keep thinking about this. 

Also, per #46 we are contemplating going bitmaps for all null handling which would make the code for integers and floats identical (outside some type promotion details maybe)